src/target/at32f43x.c

```diff
@@ -364,3 +388,150 @@ static bool at32f43_mass_erase(target_s *target)
+	for (size_t i = 0U; i < ob_count * 2U; i += 4U) {
+		const size_t offset = i >> 1U;
+		uint32_t val = target_mem32_read32(target, AT32F43x_USD_BASE + i);
+		opt_val[offset] = val & 0xffffU;
+		opt_val[offset + 1U] = val >> 16U;
+	}
+	/* Update requested entry locally */
+	opt_val[index] = value;
+	/* Wipe everything and write back. Writing matching values without an erase raises a PRGMERR. */
+	bool result = true;
+	do {
+		if (!at32f43_option_erase(target)) {
+			result = false;
+			break;
+		}
+		/* Write changed values using 16-bit accesses. */
+		for (size_t i = 0U; i < ob_count; ++i) {
+			if (!at32f43_option_write_erased(target, i, opt_val[i])) {
+				result = false;
+				break;
+			}
+		}
+	} while (false);
+
+	free(opt_val);
+	return result;
+}
+
+static bool at32f43_cmd_option(target_s *target, int argc, const char **argv)
+{
+	const uint32_t read_protected = target_mem32_read32(target, AT32F43x_FLASH_USD) & AT32F43x_FLASH_USD_RDP;
+	const bool erase_requested = argc == 2 && strcmp(argv[1], "erase") == 0;
+	/* Fast-exit if the Flash is not readable and the user didn't ask us to erase the option bytes */
+	if (read_protected && !erase_requested) {
+		tc_printf(target, "Device is Read Protected\nUse `monitor option erase` to unprotect and erase device\n");
+		return true;
+	}
+
+	/* Unprotect the option bytes so we can modify them */
+	if (!at32f43_flash_unlock(target, 0))
+		return false;
+	target_mem32_write32(target, AT32F43x_FLASH_USD_UNLOCK, AT32F43x_FLASH_KEY1);
+	target_mem32_write32(target, AT32F43x_FLASH_USD_UNLOCK, AT32F43x_FLASH_KEY2);
+
+	if (erase_requested) {
+		/* When the user asks us to erase the option bytes, kick off an erase */
+		if (!at32f43_option_erase(target))
+			return false;
+		/* Write the option bytes Flash readable key */
+		if (!at32f43_option_write_erased(target, 0U, AT32F43x_USD_RDP_KEY))
+			return false;
+	} else if (argc == 3) {
+		/* If 3 arguments are given, assume the second is an address, and the third a value */
+		const uint32_t addr = strtoul(argv[1], NULL, 0);
+		const uint32_t val = strtoul(argv[2], NULL, 0);
+		/* Try and program the new option value to the requested option byte */
+		if (!at32f43_option_write(target, addr, val))
+			return false;
+		/* Display only changes */
+		tc_printf(target, "0x%08X: 0x%04X\n", addr, val & 0xffffU);
+		return true;
+	} else
+		tc_printf(target, "usage: monitor option erase\nusage: monitor option <addr> <value>\n");
+
+	/* When all gets said and done, display the current option bytes values */
+	const target_flash_s *target_flash = target->flash;
+	const uint16_t ob_count = (target_flash->blocksize == 4096U) ? AT32F43x_4K_OB_COUNT : AT32F43x_2K_OB_COUNT;
+	for (size_t i = 0U; i < ob_count * 2U; i += 4U) {
+		const uint32_t addr = AT32F43x_USD_BASE + i;
+		const uint32_t val = target_mem32_read32(target, addr);
+		tc_printf(target, "0x%08X: 0x%04X\n", addr, val & 0xffffU);
+		tc_printf(target, "0x%08X: 0x%04X\n", addr + 2U, val >> 16U);
```

Comment:
`PRIX32` for the `uint32_t`'s here please (and same in other prints above)

Reply:
tc_printf attribute format is not merged, so I didn't notice this while copying. I thought to re-enable it locally, but then I'd get lots of noise from all the other TUs. Also I don't like the 2048 line long list, is there another way to present it?

Reply:
The next best way we can think to present that would be to have lines of 8 16-bit values, which does complicate the printing logic, but which then improves display.. in the short term, you could move the high 16 bits display into the same print as the low so that at least it halves the number of lines to 1024, which is still a lot but less.

Reply:
Block refactored, please review again. Used PRIX32 for uint32_t; as values[] are uint16_t, 0x%04X flies for this.